src/core/Sample.cpp

```diff
@@ -0,0 +1,244 @@
+		break;
+	case Loop::PingPong:
+		state->m_frameIndex = std::clamp(state->m_frameIndex, start, loopEnd);
+		if (state->m_frameIndex == loopEnd)
+		{
+			state->m_frameIndex = loopEnd - 1;
+			state->m_backwards = true;
+		}
+		else if (state->m_frameIndex <= loopStart && state->m_backwards)
+		{
+			state->m_frameIndex = loopStart;
+			state->m_backwards = false;
+		}
+		break;
+	}
+
+	playSampleRange(state, playBuffer.data(), playBuffer.size());
+
+	const auto result
+		= state->resampler().resample(&playBuffer[0][0], playBuffer.size(), &dst[0][0], numFrames, resampleRatio);
+	if (result.error != 0) { return false; }
+
+	state->m_frameIndex += (state->m_backwards ? -1 : 1) * result.inputFramesUsed;
+	amplifySampleRange(dst, result.outputFramesGenerated);
+
+	return true;
+}
+
+auto Sample::sampleDuration() const -> std::chrono::milliseconds
+{
+	const auto numFrames = endFrame() - startFrame();
+	const auto duration = numFrames / static_cast<float>(m_buffer->sampleRate()) * 1000;
+	return std::chrono::milliseconds{static_cast<int>(duration)};
+}
+
+void Sample::setAllPointFrames(int startFrame, int endFrame, int loopStartFrame, int loopEndFrame)
+{
+	setStartFrame(startFrame);
+	setEndFrame(endFrame);
+	setLoopStartFrame(loopStartFrame);
+	setLoopEndFrame(loopEndFrame);
+}
+
+void Sample::playSampleRange(PlaybackState* state, sampleFrame* dst, size_t numFrames) const
+{
+	auto framesToCopy = 0;
+	if (state->m_backwards)
+	{
+		framesToCopy = std::min<int>(state->m_frameIndex - startFrame(), numFrames);
+		copyBufferBackward(dst, state->m_frameIndex, framesToCopy);
+	}
+	else
+	{
+		framesToCopy = std::min<int>(endFrame() - state->m_frameIndex, numFrames);
+		copyBufferForward(dst, state->m_frameIndex, framesToCopy);
+	}
+
+	if (framesToCopy < numFrames) { std::fill_n(dst + framesToCopy, numFrames - framesToCopy, sampleFrame{0, 0}); }
+}
+
+void Sample::copyBufferForward(sampleFrame* dst, int initialPosition, int advanceAmount) const
+{
+	reversed() ? std::copy_n(m_buffer->rbegin() + initialPosition, advanceAmount, dst)
+			   : std::copy_n(m_buffer->begin() + initialPosition, advanceAmount, dst);
+}
+
+void Sample::copyBufferBackward(sampleFrame* dst, int initialPosition, int advanceAmount) const
+{
+	reversed() ? std::reverse_copy(
+		m_buffer->rbegin() + initialPosition - advanceAmount, m_buffer->rbegin() + initialPosition, dst)
+			   : std::reverse_copy(
+				   m_buffer->begin() + initialPosition - advanceAmount, m_buffer->begin() + initialPosition, dst);
+}
+
+auto Sample::resampleSampleRange(SRC_STATE* state, sampleFrame* src, sampleFrame* dst, size_t numInputFrames,
+	size_t numOutputFrames, double ratio) const -> SRC_DATA
+{
+	auto data = SRC_DATA{};
+	data.data_in = &src[0][0];
+	data.data_out = &dst[0][0];
+	data.input_frames = numInputFrames;
+	data.output_frames = numOutputFrames;
+	data.src_ratio = ratio;
+	data.end_of_input = 0;
+	src_process(state, &data);
+	return data;
+}
```

Comment:
Now that this has been moved into a separate class, can it be removed from here?